File: apps/libc/c/sem/sem.c
#include <sys/ipc.h>
#include <sys/sem.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <stdlib.h>
#include <errno.h>


union semun {
    int val;                    // SETVAL的值
    struct semid_ds *buf;       // IPC_STAT, IPC_SET的缓冲区
    unsigned short *array;      // GETALL, SETALL的数组
};


int sem_id;

int parent() {
    printf("父进程已启动\n");
    
    // 等待子进程启动
    printf("父进程等待子进程准备就绪...\n");
    sleep(1);
    
    // 父进程获取信号量（作为互斥锁使用）
    printf("父进程尝试获取信号量...\n");
    struct sembuf sem_op;
    sem_op.sem_num = 0;    // 操作信号量0
    sem_op.sem_op = -1;    // P操作（减1）
    sem_op.sem_flg = 0;    // 阻塞等待
    
    if (semop(sem_id, &sem_op, 1) == -1) {
        perror("父进程获取信号量失败");
        return 1;
    }
    
    printf("父进程已获取信号量，进入临界区...\n");
    
    // 模拟临界区操作
    printf("父进程在临界区工作3秒...\n");
    sleep(3);
    
    printf("父进程离开临界区，释放信号量...\n");
    
    // 释放信号量
    sem_op.sem_op = 1;     // V操作（加1）
    if (semop(sem_id, &sem_op, 1) == -1) {
        perror("父进程释放信号量失败");
        return 1;
    }
    
    printf("父进程已释放信号量\n");
    
    // 等待子进程结束
    wait(NULL);
    
    // 删除信号量集
    if (semctl(sem_id, 0, IPC_RMID) == -1) {
        perror("删除信号量集失败");
        return 1;
    }
    
    printf("父进程成功完成\n");
    return 0;
}

int child() {
    printf("子进程已启动\n");
    
    // 子进程也尝试获取信号量
    printf("子进程尝试获取信号量...\n");
    struct sembuf sem_op;
    sem_op.sem_num = 0;    // 操作信号量0
    sem_op.sem_op = -1;    // P操作（减1）
    sem_op.sem_flg = 0;    // 阻塞等待
    
    if (semop(sem_id, &sem_op, 1) == -1) {
        perror("子进程获取信号量失败");
        return 1;
    }
    
    printf("子进程已获取信号量，进入临界区...\n");
    
    // 模拟临界区操作
    printf("子进程在临界区工作2秒...\n");
    sleep(2);
    
    printf("子进程离开临界区，释放信号量...\n");
    
    // 释放信号量
    sem_op.sem_op = 1;     // V操作（加1）
    if (semop(sem_id, &sem_op, 1) == -1) {
        perror("子进程释放信号量失败");
        return 1;
    }
    
    printf("子进程已释放信号量\n");
    printf("子进程完成\n");
    return 0;
}

// 测试信号量的各种功能
void test_semaphore_features() {
    printf("\n=== 测试信号量功能 ===\n");
    
    // 创建信号量集（3个信号量）
    int test_sem_id = semget(IPC_PRIVATE, 3, IPC_CREAT | 0666);
    if (test_sem_id == -1) {
        perror("创建信号量集失败");
        return;
    }
    printf("创建了信号量集，ID为: %d\n", test_sem_id);
    
    union semun arg;
    
    // 测试1: SETVAL和GETVAL
    printf("\n测试1: SETVAL 和 GETVAL\n");
    arg.val = 5;
    if (semctl(test_sem_id, 0, SETVAL, arg) == -1) {
        perror("SETVAL 操作失败");
    } else {
        printf("将信号量0的值设置为5\n");
    }
    
    int val = semctl(test_sem_id, 0, GETVAL);
    if (val == -1) {
        perror("GETVAL 操作失败");
    } else {
        printf("获取信号量0的值: %d\n", val);
    }
    
    // 测试2: SETALL和GETALL
    printf("\n测试2: SETALL 和 GETALL\n");
    unsigned short values[3] = {10, 20, 30};
    arg.array = values;
    if (semctl(test_sem_id, 0, SETALL, arg) == -1) {
        perror("SETALL 操作失败");
    } else {
        printf("将所有信号量设置为 [10, 20, 30]\n");
    }
    
    unsigned short result[3];
    arg.array = result;
    if (semctl(test_sem_id, 0, GETALL, arg) == -1) {
        perror("GETALL 操作失败");
    } else {
        printf("获取所有信号量的值: [%d, %d, %d]\n", 
               result[0], result[1], result[2]);
    }
    
    // 测试3: 基本的semop操作
    printf("\n测试3: 基本的semop操作\n");
    struct sembuf ops[2];
    
    // 对信号量0进行P操作（减3）
    ops[0].sem_num = 0;
    ops[0].sem_op = -3;
    ops[0].sem_flg = IPC_NOWAIT;  // 非阻塞
    
    if (semop(test_sem_id, ops, 1) == -1) {
        perror("P操作失败");
    } else {
        printf("P操作成功，信号量0减少了3\n");
        val = semctl(test_sem_id, 0, GETVAL);
        printf("信号量0的当前值: %d\n", val);
    }
    
    // 对信号量1进行V操作（加5）
    ops[0].sem_num = 1;
    ops[0].sem_op = 5;
    ops[0].sem_flg = 0;
    
    if (semop(test_sem_id, ops, 1) == -1) {
        perror("V操作失败");
    } else {
        printf("V操作成功，信号量1增加了5\n");
        val = semctl(test_sem_id, 1, GETVAL);
        printf("信号量1的当前值: %d\n", val);
    }
    
    // 测试4: 多个操作的原子性
    printf("\n测试4: 多个操作的原子性\n");
    ops[0].sem_num = 0;
    ops[0].sem_op = -2;
    ops[0].sem_flg = IPC_NOWAIT;
    ops[1].sem_num = 2;
    ops[1].sem_op = -5;
    ops[1].sem_flg = IPC_NOWAIT;
    
    if (semop(test_sem_id, ops, 2) == -1) {
        printf("预期失败的原子操作（资源不足）: %s\n", 
               strerror(errno));
    } else {
        printf("原子操作成功\n");
    }
    
    // 测试5: IPC_STAT
    printf("\n测试5: IPC_STAT\n");
    struct semid_ds sem_info;
    arg.buf = &sem_info;
    if (semctl(test_sem_id, 0, IPC_STAT, arg) == -1) {
        perror("IPC_STAT 操作失败");
    } else {
        printf("信号量集信息 - 信号量数量: %lu\n", 
               sem_info.sem_nsems);
    }
    
    // 测试6: 测试错误情况
    printf("\n测试6: 错误情况测试\n");
    
    // 尝试访问不存在的信号量
    val = semctl(test_sem_id, 5, GETVAL);  // 信号量5不存在
    if (val == -1) {
        printf("预期错误，访问不存在的信号量: %s\n", 
               strerror(errno));
    }
    
    // 尝试设置超出范围的值
    arg.val = -1;  // 负值
    if (semctl(test_sem_id, 0, SETVAL, arg) == -1) {
        printf("预期错误，设置负值: %s\n", 
               strerror(errno));
    }
    
    // 清理
    if (semctl(test_sem_id, 0, IPC_RMID) == -1) {
        perror("清理失败");
    } else {
        printf("功能测试完成并清理\n");
    }
}

int main() {
    printf("=== System V 信号量测试程序 ===\n");
    
    // 首先运行功能测试
    test_semaphore_features();
    
    printf("\n=== 进程同步测试 ===\n");
    
    // 创建信号量集（1个信号量用作互斥锁）
    sem_id = semget(IPC_PRIVATE, 1, IPC_CREAT | 0666);
    if (sem_id == -1) {
        perror("创建信号量集失败");
        return 1;
    }
    
    printf("创建了信号量，ID为: %d\n", sem_id);
    
    // 初始化信号量为1（互斥锁）
    union semun arg;
    arg.val = 1;
    if (semctl(sem_id, 0, SETVAL, arg) == -1) {
        perror("初始化信号量失败");
        return 1;
    }
    printf("信号量初始化为1\n");
    
    // 创建子进程
    pid_t pid = fork();
    if (pid == -1) {
        perror("创建子进程失败");
        return 1;
    }
    
    if (pid == 0) {
        // 子进程
        return child();
    } else {
        // 父进程
        return parent();
    }
}

File: api/src/syscall/ipc/sem.rs
use alloc::vec::Vec;
use axerrno::{LinuxError, LinuxResult};
use axprocess::Pid;
use axtask::current;
use starry_core::task::TaskExt;

use crate::{
    ctypes::__kernel_time_t,
    ipc::{
        IPC_MANAGER, IPC_PRIVATE, IPC_RMID, IPC_SET, IPC_STAT, SEMMSL, SEMOPM, SemBuf, SemInfo,
        SemOpFlags,
    },
    ptr::{UserConstPtr, UserPtr, nullable},
    time::monotonic_time_nanos,
};

/// Get a semaphore set identifier
pub fn sys_semget(key: i32, nsems: i32, semflg: i32) -> LinuxResult<isize> {
    if nsems < 0 || nsems as usize > SEMMSL {
        return Err(LinuxError::EINVAL);
    }

    let cur_pid = TaskExt::from_task(&current()).thread.process().pid();
    let ipc_manager = IPC_MANAGER.lock();
    let mut sem_manager = ipc_manager.get_sem().lock();

    // If not IPC_PRIVATE, check if semaphore set already exists
    if key != IPC_PRIVATE {
        if let Some(semid) = sem_manager.get_semid_by_key(key) {
            // Existing semaphore set found
            if let Some(semset_arc) = sem_manager.get_semset_by_id(semid) {
                let semset = semset_arc.lock();
                // Check if nsems matches (if nsems > 0)
                if nsems > 0 && nsems as usize != semset.semaphores.len() {
                    return Err(LinuxError::EINVAL);
                }
                return Ok(semid as isize);
            }
        }
    }

    // Create new semaphore set
    if nsems == 0 {
        return Err(LinuxError::EINVAL);
    }

    let mode = (semflg & 0o777) as u32;
    let semid = sem_manager.create_semset(key, nsems as usize, mode, cur_pid)?;

    Ok(semid as isize)
}

/// Perform operations on semaphores
pub fn sys_semop(semid: i32, sops: UserConstPtr<SemBuf>, nsops: usize) -> LinuxResult<isize> {
    if nsops == 0 || nsops > SEMOPM {
        return Err(LinuxError::EINVAL);
    }

    // Read operations from user space
    let mut operations = Vec::with_capacity(nsops);
    for i in 0..nsops {
        let sop = sops.offset(i).get_as_ref()?;
        operations.push(*sop);
    }

    let cur_pid = TaskExt::from_task(&current()).thread.process().pid();
    let ipc_manager = IPC_MANAGER.lock();
    let sem_manager = ipc_manager.get_sem().lock();

    let semset_arc = sem_manager
        .get_semset_by_id(semid)
        .ok_or(LinuxError::EINVAL)?;

    let mut semset = semset_arc.lock();

    // Validate operations
    for op in &operations {
        if op.sem_num as usize >= semset.semaphores.len() {
            return Err(LinuxError::EFBIG);
        }
    }

    // Check if all operations can be performed immediately
    let mut should_wait = false;
    let mut has_nowait = false;

    for op in &operations {
        let flags = SemOpFlags::from_bits_truncate(op.sem_flg);
        if flags.contains(SemOpFlags::IPC_NOWAIT) {
            has_nowait = true;
        }
    }

    if !semset.can_perform_operations(&operations) {
        if has_nowait {
            return Err(LinuxError::EAGAIN);
        }
        should_wait = true;
    }

    if should_wait {
        // Add to waiting queue and block
        semset.add_waiting_process(cur_pid, operations.clone());
        let wait_queue = semset.wait_queue.clone();
        drop(semset);
        drop(sem_manager);
        drop(ipc_manager);

        // Wait for the operation to complete
        wait_queue.wait();

        // Re-acquire locks and check if operation was successful
        // let ipc_manager = IPC_MANAGER.lock();
        // let sem_manager = ipc_manager.get_sem().lock();
        // let semset_arc = sem_manager.get_semset_by_id(semid)
        //     .ok_or(LinuxError::EIDRM)?;
        // let semset = semset_arc.lock();

        // The operation should have been performed by the waking process
        // or we were woken up due to an error condition
    } else {
        // Perform operations immediately
        semset.perform_operations(&operations, cur_pid)?;

        // Add undo operations if SEM_UNDO is set
        drop(semset);
        let mut sem_manager = sem_manager;
        for op in &operations {
            let flags = SemOpFlags::from_bits_truncate(op.sem_flg);
            if flags.contains(SemOpFlags::SEM_UNDO) && op.sem_op != 0 {
                sem_manager.add_undo_operation(cur_pid, semid, op.sem_num as usize, op.sem_op);
            }
        }

        // Wake up other waiting processes
        let semset_arc = sem_manager.get_semset_by_id(semid).unwrap();
        let mut semset = semset_arc.lock();
        semset.wake_up_processes();
    }

    Ok(0)
}

// Semctl command constants
const GETVAL: u32 = 12;
const SETVAL: u32 = 16;
const GETPID: u32 = 11;
const GETNCNT: u32 = 14;
const GETZCNT: u32 = 15;
const GETALL: u32 = 13;
const SETALL: u32 = 17;

/// Control operations on semaphores
pub fn sys_semctl(semid: i32, semnum: i32, cmd: u32, arg: usize) -> LinuxResult<isize> {
    let ipc_manager = IPC_MANAGER.lock();
    let mut sem_manager = ipc_manager.get_sem().lock();

    let semset_arc = sem_manager
        .get_semset_by_id(semid)
        .ok_or(LinuxError::EINVAL)?;

    match cmd {
        IPC_RMID => {
            let mut semset = semset_arc.lock();
            semset.rmid = true;
            // Wake up all waiting processes with error
            while let Some(mut waiting) = semset.waiting_queue.lock().pop_front() {
                waiting.error = Some(LinuxError::EIDRM);
                semset.wait_queue.notify_all(true);
            }
            drop(semset);
            sem_manager.remove_semset(semid);
            Ok(0)
        }
        IPC_SET => {
            let buf_ptr = UserConstPtr::<SemInfo>::from(arg);
            let new_info = buf_ptr.get_as_ref()?;
            let mut semset = semset_arc.lock();
            semset.sem_info = *new_info;
            semset.sem_info.sem_ctime = monotonic_time_nanos() as __kernel_time_t;
            Ok(0)
        }
        IPC_STAT => {
            let buf_ptr = UserPtr::<SemInfo>::from(arg);
            let semset = semset_arc.lock();
            if let Some(buf) = nullable!(buf_ptr.get_as_mut())? {
                *buf = semset.sem_info;
            }
            Ok(0)
        }
        GETVAL => {
            if semnum < 0 || semnum as usize >= semset_arc.lock().semaphores.len() {
                return Err(LinuxError::EINVAL);
            }
            let semset = semset_arc.lock();
            Ok(semset.semaphores[semnum as usize].semval as isize)
        }
        SETVAL => {
            if semnum < 0 || semnum as usize >= semset_arc.lock().semaphores.len() {
                return Err(LinuxError::EINVAL);
            }
            let mut semset = semset_arc.lock();
            let val = arg as i16;
            if val < 0 || val as usize > crate::ipc::SEMVMX {
                return Err(LinuxError::ERANGE);
            }
            semset.semaphores[semnum as usize].semval = val;
            semset.semaphores[semnum as usize].sempid =
                TaskExt::from_task(&current()).thread.process().pid();
            semset.sem_info.sem_ctime = monotonic_time_nanos() as __kernel_time_t;
            semset.wake_up_processes();
            Ok(0)
        }
        GETPID => {
            if semnum < 0 || semnum as usize >= semset_arc.lock().semaphores.len() {
                return Err(LinuxError::EINVAL);
            }
            let semset = semset_arc.lock();
            Ok(semset.semaphores[semnum as usize].sempid as isize)
        }
        GETNCNT => {
            if semnum < 0 || semnum as usize >= semset_arc.lock().semaphores.len() {
                return Err(LinuxError::EINVAL);
            }
            let semset = semset_arc.lock();
            Ok(semset.semaphores[semnum as usize].semncnt as isize)
        }
        GETZCNT => {
            if semnum < 0 || semnum as usize >= semset_arc.lock().semaphores.len() {
                return Err(LinuxError::EINVAL);
            }
            let semset = semset_arc.lock();
            Ok(semset.semaphores[semnum as usize].semzcnt as isize)
        }
        GETALL => {
            let buf_ptr = UserPtr::<i16>::from(arg);
            let semset = semset_arc.lock();
            for (i, sem) in semset.semaphores.iter().enumerate() {
                *buf_ptr.offset(i).get_as_mut()? = sem.semval;
            }
            Ok(0)
        }
        SETALL => {
            let buf_ptr = UserConstPtr::<i16>::from(arg);
            let mut semset = semset_arc.lock();
            for (i, sem) in semset.semaphores.iter_mut().enumerate() {
                let val = *buf_ptr.offset(i).get_as_ref()?;
                if val < 0 || val as usize > crate::ipc::SEMVMX {
                    return Err(LinuxError::ERANGE);
                }
                sem.semval = val;
                sem.sempid = TaskExt::from_task(&current()).thread.process().pid();
            }
            semset.sem_info.sem_ctime = monotonic_time_nanos() as __kernel_time_t;
            semset.wake_up_processes();
            Ok(0)
        }
        _ => Err(LinuxError::EINVAL),
    }
}

/// Called when a process exits to perform undo operations
pub fn clear_proc_sem(pid: Pid) {
    let ipc_manager = IPC_MANAGER.lock();
    let mut sem_manager = ipc_manager.get_sem().lock();
    sem_manager.perform_undo_operations(pid);
}
